src/sage/schemes/elliptic_curves/ell_finite_field.py

```diff
@@ -2787,3 +2789,125 @@ def find_q(m, m4_fac, D):
                     except ValueError:
                         pass
+
+def EllipticCurve_with_prime_order(N):
+    r"""
+    Given a prime number ``N``, find another prime number `p` and construct an
+    elliptic curve `E` defined over `\mathbb F_p` such that
+    `\#E(\mathbb F_p) = N`.
+
+    INPUT:
+
+    - ``N`` -- integer; the order for which we seek an elliptic curve. Must be a
+      prime number.
+
+    OUTPUT: an elliptic curve `E/\mathbb F_p` of order ``N``
+
+    EXAMPLES::
+
+        sage: N = next_prime(int(b'sagemath'.hex(), 16))
+        sage: E = EllipticCurve_with_prime_order(N)
+        sage: E
+        Elliptic Curve defined by y^2 = x^3 + 4757897140353078952*x +
+        1841350074072114366 over Finite Field of size 8314040074357871443
+        sage: E.order() == N
+        True
+
+    The execution time largely depends on the input::
+
+        sage: N = 125577861263605878504082476745517446213
+        sage: E = EllipticCurve_with_prime_order(N) # Takes ~1 second.
+        sage: E.order() == N
+        True
+
+    TESTS::
+
+        sage: for N in prime_range(3, 100):
+        ....:     E = EllipticCurve_with_prime_order(N)
+        ....:     assert E.order() == N
+
+        sage: N = 15175980689839334471
+        sage: E = EllipticCurve_with_prime_order(N)
+        sage: E.order() == N
+        True
+
+        sage: N = next_prime(123456789)
+        sage: E = EllipticCurve_with_prime_order(N)
+        sage: E.order() == N
+        True
+
+        sage: N = 123456789
+        sage: E = EllipticCurve_with_prime_order(N)
+        Traceback (most recent call last):
+        ...
+        ValueError: input order is not prime
+
+        sage: E = EllipticCurve_with_prime_order(0)
+        Traceback (most recent call last):
+        ...
+        ValueError: input order is not prime
+
+        sage: E = EllipticCurve_with_prime_order(-7)
+        Traceback (most recent call last):
+        ...
+        ValueError: input order is not prime
+
+    .. NOTE::
+
+        Depending on the input, this function may run forever.
+
+    ALGORITHM: [BS2007]_, Algorithm 2.2
+    """
+    from sage.arith.misc import is_prime
+    from sage.combinat.subset import powerset
+    from sage.functions.other import ceil
+    from sage.misc.functional import symbolic_prod as product
+    from sage.quadratic_forms.binary_qf import BinaryQF
+    from sage.rings.fast_arith import prime_range
+    from sage.schemes.elliptic_curves.cm import hilbert_class_polynomial
+
+    if not is_prime(N):
+        raise ValueError("input order is not prime")
+
+    # The algorithm consists of multiple ‘search rounds’ for a suitable
+    # discriminant `D`, `r` defines the number of rounds. We expect this
+    # algorithm to terminate after a number of rounds that is polynomial in
+    # loglog N.
+    r = 0
+
+    while True:
+        # Iterating over the odd primes by chunks of size log(`N`).
+        S = prime_range(3, ceil((r + 1) * N.log()))
```

Comment:
If this is algorithm 2.2 in the paper, shouldn't the lower bound be `r log N`?

Also, `prime_range` rounds the inputs to the nearest integer and doesn't include the right endpoint (the rounding behaviour is a bit unintuitive in my opinion because you get things like `prime_range(2.1, 5.3)` is `[2, 3]`, but oh well).

So I think this should be:

```suggestion
        S = prime_range(max(ceil(r * N.log()), 3), floor((r + 1) * N.log()) + 1)
```

If you have any test cases where the correct answer requires the low/high point of the range to be included, those would be great to add to the TESTS in the docstring.

Reply:
Oh, I misread the paper. You need to add to S on each iteration.

Maybe do `S = prime_range(3, ceil((r + 1) * N.log()))` before the loop and `S.extend(prime_range(max(ceil(r * N.log()), 3), floor((r + 1) * N.log()) + 1))` at the end of the loop after `r += 1`?

Reply:
> shouldn't the lower bound be r log N?
Yes, and as you mentioned, you are increasing the basis of primes at each round. Though I think recreating the entire table is fine because `log N` will never be big to slow the first primes computation and since we expect `loglog N` rounds, I don't think it will make such a difference. However, to stick to the algorithm, I suggest changes in 45b053a248b5e98c7afb11a88400102595a7a00d.

> Also, prime_range rounds the inputs to the nearest integer and doesn't include the right endpoint
`ceil((r + 1) * N.log())` does just fine.

> If you have any test cases where the correct answer requires the low/high point of the range to be included, those would be great to add to the TESTS in the docstring.
This means to have D to have a prime factor of `log N`. The test
```py
sage: for N in prime_range(3, 100):
....:     E = EllipticCurve_with_prime_order(N)
....:     assert E.order() == N
```
should take care of this. It also tests the twists.

Reply:
> > Also, prime_range rounds the inputs to the nearest integer and doesn't include the right endpoint
> 
> `ceil((r + 1) * N.log())` does just fine.

When I wrote that I was thinking about the case where `(r + 1) * N.log()` is an integer, but that can't happen since `N` is an integer, so yeah the ceiling here is fine.